### DISCUSSION TITLE
Filtering index by metadata value

### DISCUSSION THREAD

**Discussion by iGeX** [ravendb/ravendb#13614] [General]
Hello

I've mentioned that I cannot filter by metadata (Raven-Clr-Type in my case) on indexing time, like this:

Map = docs =>
from doc in docs
where
(string) MetadataFor(doc)["Raven-Clr-Type"] == "some assembly qualified class name"
select new {......}

But if I put that metadata value in index, I can later filter it in query time, and it works. But I want objects of not applicable for index subclasses to be not indexed at all, but if I put above restriction in index it contains 0 entries. I've trying playing with case of class name, it gives nothing.

Is such behavior intended, or it is possible to write such index but I don't know how?

**Comment by ayende** [comment]
That should work, what happens when you try this?

**Reply by ayende** [reply]
Here is an example of us doing the same:

https://github.com/ravendb/ravendb/blob/acec81b51b21a5fe72bf3b65e4a603c496f9b1bc/test/SlowTests/Issues/RavenDB_14001.cs#L34

**Comment by iGeX** [comment]
I will include real code (tried it in raven studio directly). 
If I create indexes from code it converts it to extension methods style (poorly readable), but behavior remains the same.

1. Direct restriction

from agr in docs.DocumentModels
where (string) MetadataFor(agr)["Raven-Clr-Type"] == "modernsed.data.model.document.agreement.agreementdocumentmodel, modernsed.data"
select new
{
	BasicAgr = agr.BasicAgr.Id,
	AddAgr = agr.AddAgr.Id
}

**Result: 0 entries**

2. Restricting by range variable

from agr in docs.DocumentModels
let clrType = (string) MetadataFor(agr)["Raven-Clr-Type"]
where clrType  == "modernsed.data.model.document.agreement.agreementdocumentmodel, modernsed.data"
select new
{
	BasicAgr = agr.BasicAgr.Id,
	AddAgr = agr.AddAgr.Id
}

**Result: 0 entries**

3. No restriction, and putting metadata value to index 

from agr in docs.DocumentModels
let clrType = (string) MetadataFor(agr)["Raven-Clr-Type"]
select new
{
	BasicAgr = agr.BasicAgr.Id,
	AddAgr = agr.AddAgr.Id,
	ClrType = clrType
}

**Result: 30 465 entries**

Terms by ClrType:
**modernsed.data.model.document.agreement.agreementdocumentmodel, modernsed.data**
modernsed.data.model.document.dpp.dppdocumentmodel, modernsed.data
modernsed.data.model.document.incoming.incomingdocumentmodel, modernsed.data
modernsed.data.model.document.internal.internaldocumentmodel, modernsed.data
modernsed.data.model.document.normative.normativedocumentmodel, modernsed.data
modernsed.data.model.document.ord.orddocumentmodel, modernsed.data
modernsed.data.model.document.outgoing.outgoingdocumentmodel, modernsed.data
modernsed.data.model.document.outgoing.outgoingdummymodel, modernsed.data
modernsed.data.model.document.protocol.protocoldocumentmodel, modernsed.data
modernsed.data.model.document.universal.universaldocumentmodel, modernsed.data

So I can definitelly say that this line works and there are documents with desired ClrType in collection

let clrType = (string) MetadataFor(agr)["Raven-Clr-Type"]

But if I use it in where clause it doesn't work.

I'am using RavenDB 5.3.2 Linux version

**Reply by ayende** [reply]
You are using lower cases values, but when you index, this is using the _real_ casing.
Don't look at the terms, the terms are what happen _after_, not during the indexing. Use the proper casing for the values.
